### PR TITLE
Adding AppEvent and DeviceEvent for AMQP protocol

### PR DESCRIPTION
**Pull request by Sypheos**
referencing #325

Signed-off-by: Sylvain Prost <sylvain@thethingsindustries.com>

### CODE REVIEW

File: amqp/event.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
+package amqp
```
Comment (htdvisser) [inline amqp/event.go]: Missing copyright header

File: amqp/event.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/TheThingsNetwork/ttn/core/types"
```
Comment (htdvisser) [inline amqp/event.go]: gofmt

File: amqp/event.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+// AppEventHandler is called for events
+type AppEventHandler func(sub Subscriber, appID string, eventType types.EventType, payload []byte)
+
+// DeviceEventHandler is called for events
+type DeviceEventHandler func(sub Subscriber, appID string, devID string, eventType types.EventType, payload []byte)
+
+// PublishAppEvent publishes an event to the topic for application events of the given type
+// it will marshal the payload to json
+func (c *DefaultPublisher) PublishAppEvent(appID string, eventType types.EventType, payload interface{}) {
+	c.ctx.Info("APP EVENT")
```
Comment (htdvisser) [inline amqp/event.go]: This looks incomplete

File: amqp/event_test.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
+package amqp
+
+import (
+	"github.com/TheThingsNetwork/ttn/core/types"
+	. "github.com/TheThingsNetwork/ttn/utils/testing"
+	. "github.com/smartystreets/assertions"
+	"testing"
```
Comment (htdvisser) [inline amqp/event_test.go]: again gofmt

File: amqp/event_test.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
+package amqp
```
Comment (htdvisser) [inline amqp/event_test.go]: Again missing copyright header

File: core/handler/amqp.go [hunk 6]
```diff
@@ -66,29 +70,50 @@ func (h *handler) HandleAMQP(username, password, host, exchange, downlinkQueue s
...
 		if err != nil {
-			ctx.WithError(err).Error("Could not open publisher channel")
-			return
+			ctx.WithError(err).Warn("Could not publish Uplink")
 		}
-		defer publisher.Close()
+	}
+}
 
-		for up := range h.amqpUp {
```
Comment (htdvisser) [inline core/handler/amqp.go]: Please work with `for` and `select` instead of creating multiple goroutines

File: core/handler/downlink.go [hunk 7]
```diff
@@ -47,6 +47,8 @@ func (h *handler) EnqueueDownlink(appDownlink *types.DownlinkMessage) (err error
 					Message:        appDownlink,
 				},
 			}
+			h.mqttEvent <- t
+			//h.amqpEvent <- t
```
Comment (htdvisser) [inline core/handler/downlink.go]: Can't merge this

File: core/handler/downlink.go [hunk 8]
```diff
@@ -82,15 +84,16 @@ func (h *handler) EnqueueDownlink(appDownlink *types.DownlinkMessage) (err error
...
 		AppID: appID,
 		DevID: devID,
 		Event: types.DownlinkScheduledEvent,
 		Data: types.DownlinkEventData{
 			Message: appDownlink,
 		},
 	}
-
+	h.mqttEvent <- t
+	//h.amqpEvent <- t
```
Comment (htdvisser) [inline core/handler/downlink.go]: Can't merge this

File: core/handler/downlink.go [hunk 9]
```diff
@@ -115,6 +118,8 @@ func (h *handler) HandleDownlink(appDownlink *types.DownlinkMessage, downlink *p
 					Message:        appDownlink,
 				},
 			}
+			h.mqttEvent <- t
+			//h.amqpEvent <- t
```
Comment (htdvisser) [inline core/handler/downlink.go]: Can't merge this

File: core/handler/downlink.go [hunk 10]
```diff
@@ -192,6 +197,7 @@ func (h *handler) HandleDownlink(appDownlink *types.DownlinkMessage, downlink *p
 			Config:    downlinkConfig,
 		},
 	}
-
+	h.mqttEvent <- t
+	//h.amqpEvent <- t
```
Comment (htdvisser) [inline core/handler/downlink.go]: Can't merge this

File: core/handler/amqp.go (hunk 6 above)
Comment (Sypheos) [inline core/handler/amqp.go]: I agree, though I copied it from the handler/mqtt.go where there is two goroutine so I though I will do the same. Changing on the next commit

Comment (htdvisser) [inline core/handler/amqp.go]: Ok, didn't see that. Then let's stay consistent with the MQTT code

Comment (Sypheos) [inline core/handler/amqp.go]: Ok reverting change then

**Pull request merged**